src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -220,19 +305,149 @@ public Connection getConnection() {
         return this.subscription.connection;
     }
 
+    @Override
     public String getSubject() {
         return this.subject;
     }
 
+    @Override
     public String getReplyTo() {
         return this.replyTo;
     }
 
+    byte[] getSerializedHeader() {
+        return headers == null || headers.size() == 0 ? null : headers.getSerialized();
+    }
+
+    @Override
+    public Headers getHeaders() {
+        return headers;
+    }
+
+    @Override
     public byte[] getData() {
+        if (this.data == null)
+            return new byte[0];
         return this.data.array();
     }
 
+    @Override
+    public boolean isUtf8mode() {
+        return utf8mode;
+    }
+
+    @Override
     public Subscription getSubscription() {
         return this.subscription;
     }
+
+    @Override
+    public String toString() {
+        String hdrString = headers == null ? "" : new String(headers.getSerialized(), US_ASCII).replace("\r", "+").replace("\n", "+");
+        return "NatsMessage:" +
+                "\n  subject='" + subject + '\'' +
+                "\n  replyTo='" + replyTo + '\'' +
+                "\n  data=" + (data == null ? null : StandardCharsets.UTF_8.decode(data.asReadOnlyBuffer()).toString()) +
+                "\n  utf8mode=" + utf8mode +
+                "\n  headers=" + hdrString +
+                "\n  sid='" + sid + '\'' +
+                "\n  protocolLineLength=" + protocolLineLength +
+                "\n  protocolBytes=" + (protocolBytes == null ? null : StandardCharsets.UTF_8.decode(protocolBytes.asReadOnlyBuffer()).toString()) +
+                "\n  kind=" + kind +
+                "\n  sizeInBytes=" + sizeInBytes +
+                "\n  hdrLen=" + hdrLen +
+                "\n  dataLen=" + dataLen +
+                "\n  totLen=" + totLen +
+                "\n  subscription=" + subscription +
+                "\n  next=" + next;
+    }
+
+    /**
+     * The builder is for building normal publish/request messages,
+     * as an option for client use developers instead of the normal constructor
+     */
+    public static class Builder {

```

Comment:
You could have this class do `String` -> `ByteBuffer` conversions prior to `NatsMessage` instance initialization.